src/librustc_typeck/check/mod.rs

```diff
@@ -1311,9 +1311,40 @@ fn check_union<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     def.destructor(tcx); // force the destructor to be evaluated
     check_representable(tcx, span, def_id);
 
+    check_union_fields(tcx, span, def_id);
     check_packed(tcx, span, def_id);
 }
 
+fn check_union_fields<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                _sp: Span,
+                                item_def_id: DefId)
+                                -> bool {
+    // Without the feature we check Copy types only later
+    if !tcx.features().untagged_unions {
+        return true;
+    }
+    let t = tcx.type_of(item_def_id);
+    if let ty::Adt(def, substs) = t.sty {
+        if def.is_union() {
+            let fields = &def.non_enum_variant().fields;
+            for field in fields {
+                let field_ty = field.ty(tcx, substs);
+                // We are currently checking the type this field came from, so it must be local
+                let field_span = tcx.hir().span_if_local(field.did).unwrap();
+                let param_env = tcx.param_env(field.did);
+                if field_ty.needs_drop(tcx, param_env) {
+                    struct_span_err!(tcx.sess, field_span, E0720,
+                                     "unions may not contain fields that need dropping")
```

Comment:
Eventually, it would be nice if we could say with more detail *which type* needs to be dropped

Reply:
but I guess it doesn't have to be part of this PR